deploy/deploy.go

```diff
@@ -146,7 +162,56 @@ type kubeVersion struct {
 	ServerVersion    *kversion.Info `json:"serverVersion,omitempty" yaml:"serverVersion,omitempty"`
 }
 
+// event turns the deployment into a cluster event protobuf.
+func (d *Deployment) event() *epb.Cluster {
+	c := &epb.Cluster{}
+	switch d.Cluster.(type) {
+	case *ExternalSpec:
+		c.Cluster = epb.Cluster_CLUSTER_TYPE_EXTERNAL
+	case *KindSpec:
+		c.Cluster = epb.Cluster_CLUSTER_TYPE_KIND
+	}
+	switch d.Ingress.(type) {
+	case *MetalLBSpec:
+		c.Ingress = epb.Cluster_INGRESS_TYPE_METALLB
+	}
+	switch d.CNI.(type) {
+	case *MeshnetSpec:
+		c.Cni = epb.Cluster_CNI_TYPE_MESHNET
+	}
+	for _, cntrl := range d.Controllers {
+		switch cntrl.(type) {
```

Comment:
```
rantMode = true
defer func() { rantMode = false}()
```

While I understand why this is done and am not asking for it to be changed...

I really dislike this sort of switch.  Controllers should be able to report this themselves:

```
for _, cntrl := range d.Controllers {
    c.Controllers = append(c.Controllers, cntrl.Type())
}
```

Which would reduce the number of places that need changing when a new controller is added.  My guess is that this is not as easy because we are dealing with protobufs.

Reply:
yeah its not pretty

the other solution would be to extended the Controller interface to have a ControllerEventType() func with each spec implementing it with a method that returns the event enum. I wasn't really a fan of this solution either

Reply:
I feel better now :-)